File: app/src/main/java/com/pr0gramm/app/ui/PreviewInfo.java
package com.pr0gramm.app.ui;

import android.content.Context;
import android.graphics.drawable.Drawable;
import android.net.Uri;
import android.support.annotation.Nullable;

import com.pr0gramm.app.feed.FeedItem;
import com.pr0gramm.app.services.UriHelper;

/**
 * Info about a pixels. Can be given to a post fragment to create
 * the fragment animation on newer versions of android.
 */
public final class PreviewInfo {
    private final long itemId;
    private final int width;
    private final int height;
    private final Uri previewUri;

    @Nullable
    private final Drawable preview;


    public PreviewInfo(long id, Uri thumbnail, int width, int height, @Nullable Drawable drawable) {
        this.itemId = id;
        this.previewUri = thumbnail;
        this.width = width;
        this.height = height;
        this.preview = drawable;
    }

    public long getItemId() {
        return itemId;
    }

    @Nullable
    public Drawable getPreview() {
        return preview;
    }

    public int getWidth() {
        return width;
    }

    public int getHeight() {
        return height;
    }

    public Uri getPreviewUri() {
        return previewUri;
    }

    public static PreviewInfo of(Context context, FeedItem item) {
        return of(context, item, null);
    }

    public static PreviewInfo of(Context context, FeedItem item, @Nullable Drawable drawable) {
        Uri thumbnail = UriHelper.of(context).thumbnail(item);
        return new PreviewInfo(item.id(), thumbnail, item.width(), item.height(), drawable);
    }
}

File: app/src/main/java/com/pr0gramm/app/ui/PreviewInfo.kt
package com.pr0gramm.app.ui

import android.content.Context
import android.graphics.Bitmap
import android.graphics.drawable.Drawable
import android.net.Uri
import com.pr0gramm.app.feed.FeedItem
import com.pr0gramm.app.services.UriHelper
import com.pr0gramm.app.util.BackgroundScheduler
import com.pr0gramm.app.util.Holder
import rx.Observable

/**
 * Info about a pixels. Can be given to a post fragment to create
 * the fragment animation on newer versions of android.
 */
class PreviewInfo private constructor(val itemId: Long, val previewUri: Uri, val width: Int, val height: Int, val preview: Drawable? = null) {
    var fancy: Holder<Bitmap>? = null

    /**
     * Aspect ratio of the image this preview info represents.
     */
    val aspect: Float = width.toFloat() / height.toFloat()

    /**
     * You can pre-load a fancy thumbnail if you want to
     */
    fun preloadFancyPreviewImage(service: FancyExifThumbnailGenerator) {
        fancy = Holder.ofObservable(Observable
                .fromCallable { service.fancyThumbnail(previewUri, aspect) }
                .subscribeOn(BackgroundScheduler))
    }

    companion object {
        fun of(context: Context, item: FeedItem, drawable: Drawable? = null): PreviewInfo {
            val thumbnail = UriHelper.of(context).thumbnail(item)
            return PreviewInfo(item.id, thumbnail, item.width, item.height, drawable)
        }
    }
}
